### DISCUSSION TITLE
VBAN as in and out stream

### DISCUSSION THREAD

**Discussion by exc2k** [pschatzmann/arduino-audio-tools#1607] [Q&A] [answered]
Hey,

I want to build an full duplex communication device which sends and receives audio data over the network. I choose the VBAN protocol because it might be easy to test with the available windows and phone apps.

About the config: I set the confix to RXTX_MODE. There is one property called "stream_name". Will it be the name of the stream when I am sending audio r is it the name of the stream I want to listen to (receiving)?. I learned a lot from the many examples our are providing but unfortunately I didn't find any two way audio stream examples. I think I need two Stream-object.. one for processing the receiving and one for processing the sending audio data, right? Here is my code, would be happy for any help. Thanks so much!

```
#include <Arduino.h>
#include <AudioTools.h>
#include "AudioLibs/VBANStream.h"



AudioInfo info(44100, 1, 16);


I2SStream i2s;
VBANStream vban;

//Below only for test
//SineWaveGenerator<int16_t> sineWave( 32000);  // subclass of SoundGenerator with max amplitude of 32000
//GeneratedSoundStream<int16_t> input_stream(sineWave);  // Stream generated from sine wave


// StreamCopy(out,in)
StreamCopy copier_out(vban, i2s); // Copy from i2s-microfone to VBAN 
StreamCopy copier_in(i2s, vban, 2048); // Copy from VBAN to i2s-speaker (44100Hz needs 2048 byte of buffer)


// Arduino Setup
void setup(void) {  

  // Open Serial 
  Serial.begin(115200);

  // change to Warning to improve the quality
  AudioLogger::instance().begin(Serial, AudioLogger::Info); 


  // setup input from vban
  Serial.println("starting VBAN...");

  auto cfg_vban = vban.defaultConfig(RXTX_MODE);
  cfg_vban.ssid = "test_wifi";
  cfg_vban.password = "supersecret";
  cfg_vban.stream_name = "Talkie"; //name of stream
  
  cfg_vban.target_ip = IPAddress{192,168,1,10}; 

  //cfg_vban.throttle_active = true;  //needs to be true when using sinewave generator
  //cfg.throttle_correction_us = 0; // optimize overload and underrun

  vban.begin(cfg_vban);

  Serial.println("starting VBAN...");


  // start I2S in and out
  Serial.println("starting I2S...");

  auto cfg_i2s = i2s.defaultConfig(RXTX_MODE);

  cfg_i2s.copyFrom(info); 
  cfg_i2s.i2s_format = I2S_STD_FORMAT;
  cfg_i2s.pin_ws = 14;
  cfg_i2s.pin_bck = 15;
  cfg_i2s.pin_data = 18;
  cfg_i2s.pin_data_rx = 19;
  //cfg_i2s.fixed_mclk = sample_rate * 256;
  //cfg_i2s.pin_mck = 3; // must be 0,1 or 3 - only for ESP_IDF_VERSION_MAJOR >= 4

  i2s.begin(cfg_i2s);

  Serial.println("I2S started...");
}


// Arduino loop - copy sound to out 
void loop() {
  copier_in.copy();
  copier_out.copy();
}
```

**Comment by pschatzmann** [comment, marked as answer]
I don't know since I have never tried this. 

If this is not  working with 2 separate instances, feel free to extend the implementation of [the VBANStream class](https://github.com/pschatzmann/arduino-audio-tools/blob/main/src/AudioLibs/VBANStream.h)

**Reply by exc2k** [reply]
Thanks for your reply. I will try with two instances, one for in and one for out. But just to understand how to realize full duplex at all, can you confim the following sketch:

1.) I need 2 in and 2 out stream objects (as far as the class does not support in and out at the same time like the i2s class). I use I2S and VBAN as example here but it can be every other Stream (like UDP)...

```
I2SStream i2s_in;
I2SStream i2s_out;
VBANStream vban_in;
VBANStream vban_out;
```

2.) I need two StreamCopier like this:

```
StreamCopy copier_out(vban_out, i2s_in); // Copy from is2-microfone to VBAN 
StreamCopy copier_in(i2s_out, vban_in, 2048); // Copy from VBAN to i2s-speaker (44100Hz needs 2048 byte of buffer)
```

3.) I configurate everything in the void setup function

4.) I copy both streams in the main loop

```
void loop() {
  copier_in.copy();
  copier_out.copy();
}
```

- Can you confirm this sketch? Have you ever tried any kind of full duplex application or do you have any example?

Vielen Dank :)

**Reply by pschatzmann** [reply]
Yes, that's what I would try first.
